### PR TITLE
Add `format` and `cformat` modules from `RustPython`

### PR DESCRIPTION
**Pull request by MichaReiser**
This PR moves the `format` and `cformat` modules from` RustPython` into the new  `format` crate. I'm not sure if this is the right crate to add. Please let me know if I should move it to another crate and/or add it behind a feature flag. 

The `format` implementation has an optimized path for `BorrowedStr`. I kept that path by introducing a new `CharLen` trait that `BorrowedStr` can implement in `RustPython`. 

The goal of this is that `format` and `cformat` are the last two modules from `RustPython` that ruff depends on. Moving them into this repository would allow us to cut that dependency.

### CODE REVIEW

File: literal/src/format.rs [hunk 1]
```diff
@@ -1,5 +1,1229 @@
...
+        self.inner.len()
+    }
+}
+
+impl Deref for AsciiStr<'_> {
+    type Target = str;
+    fn deref(&self) -> &Self::Target {
+        self.inner
+    }
+}
```
Comment (MichaReiser) [inline literal/src/format.rs:685]: This is new

**Comment by youknowone** [comment]
I feel a bit awkward to put format functions to parser directory. But that will be fine if it helps to maintaining Ruff in practice.
rustpython-common has too many dependencies to handling platform/threading stuff.
To make future contributors not being confused, could you add comments to format.rs and cformat.rs that it doesn't look like parser component but added to support common parser user scenario?

**Comment by MichaReiser** [comment]
> I feel a bit awkward to put format functions to parser directory. But that will be fine if it helps to maintaining Ruff in practice. rustpython-common has too many dependencies to handling platform/threading stuff. To make future contributors not being confused, could you add comments to format.rs and cformat.rs that it doesn't look like parser component but added to support common parser user scenario?

I understand. Would it help if we move the logic into a new `rustpython-format` crate?

**Comment by MichaReiser** [comment]
I added a new crate and updated the repository URLs.

**Review by youknowone — APPROVED** [review]

**Comment by youknowone** [comment]
Thank you!

**Review by MichaReiser — COMMENTED** [review]
a comment to make GitHub happy 🤔

**Pull request merged**